### ISSUE TITLE
`cargo lambda build` fails with `failed to start 'cargo metadata': Permission denied (os error 13)` error

### ISSUE THREAD

**Issue by ryanblock** [cargo-lambda/cargo-lambda#420]
Following the [directions here](https://www.cargo-lambda.info/guide/installation.html) I have both installed Cargo Lambda via `pip3 install cargo-lambda` and manually installed it (and Zig) to my machine following normal procedures (e.g. a folder in my path with correct permissions, etc.).

Both scenarios produce the exact same result: when running any `cargo lambda build` commands, I always get the error: `failed to start 'cargo metadata': Permission denied (os error 13)` + `Permission denied (os error 13)`. This error can only be worked around by running as sudo (which we don't want to do for all the obvious reasons).

Running `cargo metadata` (and other cargo commands) work fine; running `cargo build` in the same folder also works fine. Only `cargo lambda build` (etc.) exhibits this behavior.

See terminal below:

```sh
$ zig version
0.11.0-dev.2298+5d63d1115

$ cargo lambda
Cargo Lambda is a CLI to work with AWS Lambda functions locally

Usage: cargo lambda [OPTIONS] [COMMAND]...

$ cargo lambda build --release
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)

$ cargo metadata
warning: please specify `--format-version` flag explicitly to avoid compatibility problems
{"packages"...

$ cargo build --release
       Fresh unicode-ident v1.0.8...
```

Environment: M2 Mac + latest Ventura, bash

**Comment by ryanblock** [comment]
Sure thing, tried that release and unfortunately no luck. The version looks correct, that hash is the latest commit in the repo (`3bab2a5... Add extra verification of project directory. (#423)`):

```sh
$ cargo lambda --version
cargo-lambda 0.18.1 (3bab2a5 2023-03-31Z)

$ RUST_LOG=trace cargo lambda build --release
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: false, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: true, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 0, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)

$ ls
total 64
drwxr-xr-x  7 ryan  staff   224B Mar 29 09:09 ./
drwxr-xr-x  3 ryan  staff    96B Mar 29 08:50 ../
-rw-r--r--  1 ryan  staff     7B Mar 29 08:50 .gitignore
-rw-r--r--  1 ryan  staff    22K Mar 29 08:51 Cargo.lock
-rw-r--r--  1 ryan  staff   776B Mar 29 08:50 Cargo.toml
drwxr-xr-x  3 ryan  staff    96B Mar 29 08:50 src/
```

**Comment by ryanblock** [comment]
Not sure I follow, what permissions would you like me to demonstrate that are not presented in my last comment? The above `ls` shows the current and parent directory are `drwxr-xr-x` by my user (`ryan`). The folder and files in question only exist because my user ran `cargo lambda new bootstrap` to generate it, and are all `-rw-r--r--`. Either way, here's a video of a run of creating an entirely new project and it failing to build with the permissions error:

https://user-images.githubusercontent.com/200964/229188559-3e615d8a-ebd4-472b-9e4c-7e78df276d16.mov

At the end, apologies for the `CARGO_LOG` typo, meant to do `RUST_LOG`. The output is the same as above, but presented here (again) for reference:
```sh
⮑  CARGO_LOG=trace cargo lambda build --release
[2023-03-31T17:19:45Z TRACE cargo::util::config] get cv ConfigKey { env: "CARGO_ALIAS_LAMBDA", parts: [("alias", 5), ("lambda", 11)] }
[2023-03-31T17:19:45Z TRACE cargo::util::config] get cv ConfigKey { env: "CARGO_TERM", parts: [("term", 5)] }
[2023-03-31T17:19:45Z TRACE cargo::util::config] get cv ConfigKey { env: "CARGO_NET", parts: [("net", 5)] }
[2023-03-31T17:19:45Z TRACE cargo::util::config] get cv ConfigKey { env: "CARGO_HTTP", parts: [("http", 5)] }
Error:
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)

× [10:19:45] (╯ಠ益ಠ)╯︵~/git/lambda-cargo-test/bootstrap
⮑  RUST_LOG=trace cargo lambda build --release
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: false, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: true, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 0, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }
Error:
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)
```

**Comment by ryanblock** [comment]
```sh
$ sudo cargo clean
Password:

$ cargo metadata --no-deps --manifest-path Cargo.toml
warning: please specify `--format-version` flag explicitly to avoid compatibility problems
{"packages":[{"name":"bootstrap","version":"0.1.0","id":"bootstrap 0.1.0 (path+file:///Users/ryan/git/lambda-cargo-test/bootstrap)","license":null,"license_file":null,"description":null,"source":null,"dependencies":[{"name":"lambda_http","source":"registry+https://github.com/rust-lang/crates.io-index","req":"^0.7","kind":null,"rename":null,"optional":false,"uses_default_features":false,"features":["apigw_http"],"target":null,"registry":null},{"name":"lambda_runtime","source":"registry+https://github.com/rust-lang/crates.io-index","req":"^0.7","kind":null,"rename":null,"optional":false,"uses_default_features":true,"features":[],"target":null,"registry":null},{"name":"tokio","source":"registry+https://github.com/rust-lang/crates.io-index","req":"^1","kind":null,"rename":null,"optional":false,"uses_default_features":true,"features":["macros"],"target":null,"registry":null},{"name":"tracing","source":"registry+https://github.com/rust-lang/crates.io-index","req":"^0.1","kind":null,"rename":null,"optional":false,"uses_default_features":true,"features":["log"],"target":null,"registry":null},{"name":"tracing-subscriber","source":"registry+https://github.com/rust-lang/crates.io-index","req":"^0.3","kind":null,"rename":null,"optional":false,"uses_default_features":false,"features":["fmt"],"target":null,"registry":null}],"targets":[{"kind":["bin"],"crate_types":["bin"],"name":"bootstrap","src_path":"/Users/ryan/git/lambda-cargo-test/bootstrap/src/main.rs","edition":"2021","doc":true,"doctest":false,"test":true}],"features":{},"manifest_path":"/Users/ryan/git/lambda-cargo-test/bootstrap/Cargo.toml","metadata":null,"publish":null,"authors":[],"categories":[],"keywords":[],"readme":null,"repository":null,"homepage":null,"documentation":null,"edition":"2021","links":null,"default_run":null,"rust_version":null}],"workspace_members":["bootstrap 0.1.0 (path+file:///Users/ryan/git/lambda-cargo-test/bootstrap)"],"resolve":null,"target_directory":"/Users/ryan/git/lambda-cargo-test/bootstrap/target","version":1,"workspace_root":"/Users/ryan/git/lambda-cargo-test/bootstrap","metadata":null}

$ RUST_BACKTRACE=1 cargo lambda build --release
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)
```

**Comment by ryanblock** [comment]
Anything else I can do to help here? Happy to jump on a screen share to further debug.

In the mean time, is the workaround to just run a normal `cargo build --release` from a Linux machine?

**Comment by ryanblock** [comment]
I didn't dig super deep but on a quick pass I didn't see any tests that verify cutting a release is ever run inside the macOS build environment – since you don't run macOS, perhaps it's worth checking to see that it is indeed run?

**Comment by ryanblock** [comment]
Ah, dope! Ok, this inspired me to keep sciencing and verify that tests pass here, which they do not:
<pre>
$ cargo test --all-features
[build output]

     Running unittests src/main.rs (target/debug/deps/cargo_lambda-b834993371c97e24)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/cargo_lambda.rs (target/debug/deps/cargo_lambda-88bcf24ca669a61e)

running 7 tests
test test_init_subcommand_without_override ... ok
test test_build_http_function ... FAILED
test test_build_basic_function ... FAILED
test test_build_telemetry_extension ... FAILED
test test_build_logs_extension ... FAILED
test test_build_basic_extension ... FAILED
test test_init_subcommand ... FAILED

failures:

---- test_build_http_function stdout ----
thread 'test_build_http_function' panicked at 'Expected success, was 1
stdout:
```
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: false, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: false, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 2, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }

```
stderr:
```
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)


```
', crates/cargo-lambda-cli/tests/cargo_lambda.rs:39:49
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- test_build_basic_function stdout ----
thread 'test_build_basic_function' panicked at 'Expected success, was 1
stdout:
```
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: false, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: false, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 2, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }

```
stderr:
```
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)


```
', crates/cargo-lambda-cli/tests/cargo_lambda.rs:22:49

---- test_build_telemetry_extension stdout ----
thread 'test_build_telemetry_extension' panicked at 'Expected success, was 1
stdout:
```
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: true, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: false, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 2, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }

```
stderr:
```
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)


```
', crates/cargo-lambda-cli/tests/cargo_lambda.rs:98:10

---- test_build_logs_extension stdout ----
thread 'test_build_logs_extension' panicked at 'Expected success, was 1
stdout:
```
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: true, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: false, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 2, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }

```
stderr:
```
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)


```
', crates/cargo-lambda-cli/tests/cargo_lambda.rs:78:10

---- test_build_basic_extension stdout ----
thread 'test_build_basic_extension' panicked at 'Expected success, was 1
stdout:
```
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: true, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: false, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 2, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }

```
stderr:
```
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)


```
', crates/cargo-lambda-cli/tests/cargo_lambda.rs:58:10

---- test_init_subcommand stdout ----
thread 'test_init_subcommand' panicked at 'Expected success, was 1
stdout:
```
TRACE run: building project options=Build { output_format: Binary, lambda_dir: None, arm64: false, x86_64: false, extension: false, flatten: None, compiler: CargoZigbuild, build: Build { common: CommonOptions { quiet: false, jobs: None, keep_going: false, release: false, profile: None, features: [], all_features: false, no_default_features: false, target: [], target_dir: None, manifest_path: None, ignore_rust_version: false, message_format: [], unit_graph: false, verbose: 2, color: None, frozen: false, locked: false, offline: false, config: [], unstable_flags: [], timings: None }, packages: [], workspace: false, exclude: [], all: false, lib: false, bin: [], bins: false, example: [], examples: false, test: [], tests: false, bench: [], benches: false, all_targets: false, out_dir: None, build_plan: false, future_incompat_report: false } }
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading Cargo metadata
TRACE run:load_metadata{manifest_path="Cargo.toml"}: loading cargo metadata metadata=MetadataCommand { cargo_path: None, manifest_path: Some("Cargo.toml"), current_dir: None, no_deps: true, features: [], all_features: false, no_default_features: false, other_options: [], env: {}, verbose: true }

```
stderr:
```
Error: 
  × failed to start `cargo metadata`: Permission denied (os error 13)
  ╰─▶ Permission denied (os error 13)


```
', crates/cargo-lambda-cli/tests/cargo_lambda.rs:116:49


failures:
    test_build_basic_extension
    test_build_basic_function
    test_build_http_function
    test_build_logs_extension
    test_build_telemetry_extension
    test_init_subcommand

test result: FAILED. 1 passed; 6 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.83s

error: test failed, to rerun pass `-p cargo-lambda --test cargo_lambda`
<pre>

Anything catch your eye in there, by chance?

**Comment by ryanblock** [comment]
Ok, I found and was able to fix the issue! Given what you demonstrated above, I wasn't able to replicate the issue on two other Macs I had in my office, so I knew this was isolated to my individual machine.

I went looking for how `cargo metadata` was called within `cargo lambda`, because clearly invoking `cargo metadata` by itself was not the issue, specifically.

So it turns out that `cargo metadata` is called via the [`cargo_metadata` crate](https://github.com/oli-obk/cargo_metadata), which has its own logic for how to execute `cargo metadata` in a child process. That includes looking for and using `CARGO` env var: https://github.com/oli-obk/cargo_metadata/blob/main/src/lib.rs#L754

On my machine, `CARGO` was set to a value that was unusable by this crate for this purpose. That is the source of the `Permission denied (os error 13)` error; knowing the cause of the issue, it's disappointingly vague and difficult to trace.

So as the saying goes: To dependencies! The cause of and the solution to all of life's problems.

Thanks for your patience in helping me work through this, I'm now unblocked.
